### ISSUE TITLE
Navmesh generation problems in 0.54

### ISSUE THREAD

**Issue by sweet235** [Unvanquished/Unvanquished#2403]
I'll try to collect all problems with the ingame navmesh generator here. Maybe I'll split this into several issues later.

- [x] Use BSP Patches
  * https://github.com/Unvanquished/Unvanquished/issues/2389
  * https://github.com/Unvanquished/Unvanquished/pull/2402
- [ ] RC_MAX_LAYERS is hit
- [x] Tweaking of cell height would help
  * https://github.com/Unvanquished/Unvanquished/pull/2407
- [ ] Only partial navmeshes
- [x] Navmesh generation can take a long time
- [x] A local game seems to create navmeshes twice

### Use BSP Patches

Navgen should be blocked by curvy objects.

### RC_MAX_LAYERS is hit
In the past, some maps required increasing `RC_MAX_LAYERS` in `recastnavigation/Recast/Source/RecastLayers.cpp` to successfully create navmeshes. This "fix" suggested an investigation, which I never performed. But it did give us navmeshes that seemed to work.

One example for this is [map-transparency](https://users.unvanquished.net/~sweet/pkg/map-transparency_0%2Balpha2.dpk). I observe this when loading it:

```Generating navmesh for level0 
Generating navmesh for level1 
Generating navmesh for level2 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level2upg 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level3 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
Generating navmesh for human_bsuit
```

The old navmeshes for this map were created using this in `RecastLayers.cpp`:
```
-static const int RC_MAX_LAYERS = 63;
+static const int RC_MAX_LAYERS = 255;
```

Other maps with this problem are: [map-hq-beta28](https://users.unvanquished.net/~sweet/pkg/map-hq-beta28_0%2B0.dpk), [map-arkadia](https://users.unvanquished.net/~sweet/pkg/map-arkadia_0%2B0.dpk).

### Tweaking of cell height would help
The (gigantic) [map-anthill](https://users.unvanquished.net/~sweet/pkg/map-anthill_1.dpk) gives this:

```
Generating navmesh for level0 
Warn: Navmesh generation for Dretch failed: Map is too tall to generate a navigation mesh
```

In the past, I tweaked this line: https://github.com/DaemonEngine/daemonmap/blob/master/tools/quake3/q3map2/nav.cpp#L1086, using a value the near printf's suggested (I used `height = 18;`). That convinced daemonmap to create navmeshes. I think this could be done by command line arguments in daemonmap, but I'm not sure.

It is unclear how we should give command line options to the ingame navmesh generator.

### Only partial navmeshes

The (gigantic) [map-foggy](https://users.unvanquished.net/~sweet/pkg/map-foggy_0%2Bb1.dpk) receives only partial navmeshes. This seems to be related to #2357, ~~but i cannot even post a screenshot~~. The game dies when i try to load this map in a local game for `/navedit`.

~~But I can create navmeshes on a server and connect a client. Then I see that the bots freeze at `(-13000, 13000, 20)`. But they run around at other places.~~

Screenshot:
![unvanquished_2023-02-01_124834_000](https://user-images.githubusercontent.com/110394607/216035127-77615798-ef25-41c1-9dc0-0b1c92b41dbf.jpg)

Screenshot was created by generating navmeshes in daemonded, then loading them into a local daemon game.

This is not a new problem, daemonmap did the same.

### Navmesh generation can take a long time

For map-foggy, you need a rather beefy server to create the navmeshes without a delay of more than a minute. Even worse, my experiments suggest that during the navmesh generation phase, the server might not provide download links to clients. So users see a frozen download screen at 0%.

###  A local game seems to create navmeshes twice

When doing `$ ./daemon -pakpath ~/.local/share/unvanquished/base/pkg +devmap plat23`, I observe this:

```
Generating navmesh for level1 
Generating navmesh for level2 
Generating navmesh for level2upg 
Generating navmesh for level3 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
Generating navmesh for human_bsuit 
glsl shaders took 59 msec to build 
Sweet.genius entered the game 
Generating navmesh for level2 
Generating navmesh for level2upg 
Generating navmesh for level3 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
Generating navmesh for human_bsuit 
==== Bot Navigation Initialization ==== 
 loading navigation mesh file 'maps/plat23-level0.navMesh'... 
...
```

**Comment by slipher** [comment]
> Even worse, my experiments suggest that during the navmesh generation phase, the server might not provide download links to clients. So users see a frozen download screen at 0%.

If you use the *automatic* navgen generation, the server is expected to remain responsive during generation. The `navgen` command on the other hand blocks the main thread. Do you have `navgen` commands in your map config?

> A local game seems to create navmeshes twice

I'm suspecting something in the config again as sgame's automatic navgen only triggers after the  `==== Bot Navigation Initialization ====` line.

**Comment by slipher** [comment]
> Tweaking of cell height would help

Should be fixed by #2407 (automatically sets the cell height)

> RC_MAX_LAYERS is hit

I wonder if increasing the cell height could help here as well. It would need to be manually increased using #2419

**Comment by bmorel** [comment]
```
Unvanquished/libs/recastnavigation% grep --exclude-dir=.git 'RC_MAX_LAYERS =' -r -B2
Recast/Source/RecastLayers.cpp-// Must be 255 or smaller (not 256) because layer IDs are stored as
Recast/Source/RecastLayers.cpp-// a byte where 255 is a special value.
Recast/Source/RecastLayers.cpp:static const int RC_MAX_LAYERS = 63;
```

FYI in my daemonmap I used 255 (max value) without any negative consequence I could see. This is used to statically allocate a memory array, and I doubt you want to change recast's code enough to make this dynamic.

**Comment by illwieckz** [comment]
Interestingly, the `foggy` map prints this log, exactly like the `anthill` one, saying there is no lightmap and no lightgrid:

```
Warn: no lightmap files found 
Warn: light grid mismatch, default light grid used
```

The levelshot states the map was also done by Seeeker, like `anthill`.

So I assume this map was done using Blender, converted as a model, then the model was inserted in a “large room” in a `.map` and the `.map` file compiled to `.bsp`. I'm not surprised the map hits some corner cases… 😁️

**Comment by bmorel** [comment]
It seems that some cannon maps are also affected by navmesh generation failures, like rotcannon-garbage. I doubt this is about blender.
Logs:

```
Existing navmesh file maps/rotcannon-garbage-level0.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level1.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level2.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level2upg.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level3.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level3upg.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-level4.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-human_naked.navMesh can't be used: File is wrong version 
Existing navmesh file maps/rotcannon-garbage-human_bsuit.navMesh can't be used: File is wrong version 
Generating navmesh for level0 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level1 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level2 
Generating navmesh for level2upg 
Generating navmesh for level3 
Generating navmesh for level3upg 
Generating navmesh for level4 
Generating navmesh for human_naked 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for human_bsuit 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
glsl shaders took 544 msec to build 
[bot] Bot#1 disconnected 
[bot] Bot#2 disconnected 
[bot] Bot#3 disconnected 
[bot] Bot#4 disconnected 
[bot] Bot#5 disconnected 
[bot] Bot#6 disconnected 
[bot] Bot#7 disconnected 
[bot] Bot#8 disconnected 
[bot] Bot#9 disconnected 
[bot] Bot#10 disconnected 
freem entered the game 
]/bot fill 5 
==== Bot Navigation Initialization ==== 
 loading navigation mesh file 'maps/rotcannon-garbage-level0.navMesh'... 
Warn: Loading navmesh for level0 failed: File is wrong version 
 loading navigation mesh file 'maps/rotcannon-garbage-level1.navMesh'... 
Warn: Loading navmesh for level1 failed: File is wrong version 
 loading navigation mesh file 'maps/rotcannon-garbage-level2.navMesh'... 
 loading navigation mesh file 'maps/rotcannon-garbage-level2upg.navMesh'... 
 loading navigation mesh file 'maps/rotcannon-garbage-level3.navMesh'... 
 loading navigation mesh file 'maps/rotcannon-garbage-level3upg.navMesh'... 
 loading navigation mesh file 'maps/rotcannon-garbage-level4.navMesh'... 
 loading navigation mesh file 'maps/rotcannon-garbage-human_naked.navMesh'... 
Warn: Loading navmesh for human_naked failed: File is wrong version 
 loading navigation mesh file 'maps/rotcannon-garbage-human_bsuit.navMesh'... 
Warn: Loading navmesh for human_bsuit failed: File is wrong version 
Generating navmesh for level0 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for level1 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for human_naked 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
Generating navmesh for human_bsuit 
rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. 
==== Bot Navigation Initialization ==== 
 loading navigation mesh file 'maps/rotcannon-garbage-level0.navMesh'... [further messages like this will be suppressed] 
Warn: Loading navmesh for level0 failed: File is wrong version 
Warn: Cannot load navigation mesh file for level0 
[bot] Bot#1 is connecting… 
[bot] Bot#1 entered the game 
[bot] Bot#2 is connecting… 
[bot] Bot#2 entered the game 
[bot] Bot#3 is connecting… 
[bot] Bot#3 entered the game 
[bot] Bot#4 is connecting… 
[bot] Bot#4 entered the game 
[bot] Bot#5 is connecting… 
[bot] Bot#5 entered the game 
[bot] Bot#6 is connecting… 
[bot] Bot#6 entered the game 
[bot] Bot#7 is connecting… 
[bot] Bot#7 entered the game 
[bot] Bot#8 is connecting… 
[bot] Bot#8 entered the game 
[bot] Bot#9 is connecting… 
[bot] Bot#9 entered the game 
[bot] Bot#10 is connecting… 
[bot] Bot#10 entered the game 
Server: generating bot navigation mesh for level0... 0% 
Bot navmesh generation failed! 
[bot] Bot#1 disconnected 
[bot] Bot#2 disconnected 
[bot] Bot#3 disconnected 
[bot] Bot#4 disconnected 
[bot] Bot#5 disconnected 
[bot] Bot#6 disconnected 
[bot] Bot#7 disconnected 
[bot] Bot#8 disconnected 
[bot] Bot#9 disconnected 
[bot] Bot#10 disconnected 
```

I wonder what's all this about "wrong version". In my opinion, a navmesh version should really only be incompatible with a game or mod if either:

* the map geometry changed, this can be checked by a checksum of the map, I guess;
* the agents properties change, and there you would not want to remove old navmeshes, as this might simply be about using a mod, and generating those is slow, and yes, it should be expected that server owners will host mods with different agent properties, to test changes for example;
* some critical generator properties were changed (including important bugfixes), and I'm not even certain this one is good enough reason to make version incompatible when it's impossible to generate new navmeshes.

**Comment by illwieckz** [comment]
@sweet235 I edited your post to add a list of check boxes, I also listed the patch stuff (and marked it as done).

**Comment by bmorel** [comment]
> Navmesh generation can take a long time

This can't really be fixed, navmesh generation will always depends on the size of the map. It can be greatly improved though, **if** several navmeshes are generated in parallel, and if the machine have several virtual cores. I wonder how many people rent a multi-core servers to host their mod though.

Other ways to really improve it, which would require actually useful changes, would be:

* no longer generate navmeshes outside of reachable map area. Problem being: how could one automatically know where is this area? The only way I see for that is to consider stuff that is not reachable from a spawn point as unreachable;
* partial generation navmeshes, that is, only build navmeshes around bots. OpenMW do that, except it's around the player obviously. This would be tricky though, and it would have the interesting consequence that bots would naturally "explore" their environment at map start, as they'll have to roam in tiny parts of the map. I'm not sure it's worth it though, as this might be very complex to do correctly, and unlike OpenMW which have to support really huge maps, unvanquished maps are rather small (I mean, it's not the same order of magnitude **at all**);

**Comment by sweet235** [comment]
> Do you have navgen commands in your map config?

No. At least none that I can find.

> I edited your post to add a list of check boxes

Thanks.

**Comment by sweet235** [comment]
The doubled navmesh generation goes away when I remove my `map/default.cfg`. But it happens when I have it present, with this single line in it:

```
bot fill 0
```

**Comment by bmorel** [comment]
Since the element: ￼ "Only partial navmeshes" is not a regression, I guess it can be ignored for now? I think there is also a fix for RC_MAX_LAYERS in master, but I can not say for certain.

**Comment by slipher** [comment]
Upon merging #2419 we will have parity with daemonmap so it would make sense to close this. We can open a new issue for RC_MAX_LAYERS (I don't believe there has been a fix. Partial navmeshes already seems to have its own issue.

**Comment by bmorel** [comment]
I do not think daemonmap had a per-map configuration file. I did wrote a patch to it that I've used a lot to have some stuff configurable without recompilation, but I think this problem is also solved by the merging into gamelogic (bbox are configured here).

The main point of the merging in gamelogic was to use the gamelogic's configuration for agents to me, which is something that was previously hardcoded in daemonmap. To be on par with daemonmap, we'd need parallel generation of navmeshes, but I don't think this is going to happen soon anyway. Maybe if it was possible to summon daemon to only generate navmesh for a single class, this could be used to work around that?
In any case, it's an optimisation issue, not a feature or stability problem, so I'm fine with the situation.
Actually, I'm thinking that currently the code might be finally mature enough for me to port the (reliable) movement improvements I did with betterai mod: crouching and lesser-auto-jump. I'm not sure if I should wait for a 0.54.1 before doing that work or not, though, as I'm still concerned with the very poor state of the stable release, and the problems it triggers to people who try to host a server (and let's not talk about hosting a mod!)

**Comment by slipher** [comment]
> I do not think daemonmap had a per-map configuration file.

#2419 is meant to replace command-line flags in daemonmap.

**Comment by bmorel** [comment]
I see. Well, I've played a bit with daemonmap as you know, and I never used those parameters, notably because I was unable to understand what they implied.
The only thing I had to tweak so far was a security threshold for a feature (lesser-autojump) which is *absent* from vanilla daemonmap or current gamelogic. It's also a per-map feature more than a per-agent one. Talk about this  likely belongs to a PR adding the feature.
I would consider the options of daemonmap as deadcode, until I'm proven wrong. As far as I know, nobody around know what they really are for, and they easily might be just for experimenting with daemonmap itself when it was maintained.

**Comment by bmorel** [comment]
About `RC_MAX_LAYERS` , I think one of the problems here is that unv (and daemonmap before) generates navmeshes on places which are not walkable, such as  chasm's sky (see #1385). I suppose that getting rid of those would reduce the RC_MAX_LAYER problem, as it says: `rcBuildHeightfieldLayers: layer overflow (too many overlapping walkable platforms). Try increasing RC_MAX_LAYERS. `
Maps with many layers that really are walkable will always have a problem with this, though, as long as the parameter is not made dynamic, and even if made dynamic, 255 is a rather "small" number so things can still be troublesome for extreme situations. Fortunately, so far, none of the 300+ ported maps requires higher number...

**Comment by bmorel** [comment]
I'd like to point at the fact that daemonmap was able to generate navmeshes for sokolov in the past, so we still have regressions around.
